fwdctl/src/dump.rs

```diff
@@ -24,50 +28,230 @@ pub struct Options {
+    )]
+    pub output_format: String,
+
     #[arg(short = 'x', long, help = "Print the keys and values in hex format.")]
     pub hex: bool,
 }
 
 pub(super) async fn run(opts: &Options) -> Result<(), api::Error> {
     log::debug!("dump database {:?}", opts);
-    let cfg = DbConfig::builder().truncate(false);
 
+    let cfg = DbConfig::builder().truncate(false);
     let db = Db::new(opts.db.clone(), cfg.build()).await?;
     let latest_hash = db.root_hash().await?;
     let Some(latest_hash) = latest_hash else {
         println!("Database is empty");
         return Ok(());
     };
     let latest_rev = db.revision(latest_hash).await?;
-    let start_key = opts.start_key.clone().unwrap_or(Box::new([]));
+
+    let start_key = opts.start_key.clone().unwrap_or_else(|| Box::new([]));
+    let stop_key = opts.stop_key.clone().unwrap_or_else(|| Box::new([]));
+    let mut key_count = 0;
+
     let mut stream = MerkleKeyValueStream::from_key(&latest_rev, start_key);
-    loop {
-        match stream.next().await {
-            None => break,
-            Some(Ok((key, value))) => {
-                if opts.hex {
-                    println!("'{}': '{}'", hex::encode(&key), hex::encode(&value));
-                } else {
-                    println!("'{}': '{}'", u8_to_string(&key), u8_to_string(&value));
+    let mut output_handler = create_output_handler(opts).expect("Error creating output handler");
+
+    while let Some(item) = stream.next().await {
+        match item {
+            Ok((key, value)) => {
+                output_handler.handle_record(&key, &value);
+
+                key_count += 1;
+
+                if key == stop_key || key_count_exceeded(opts.max_key_count, key_count) {
+                    handle_next_key(stream.next().await).await;
+                    break;
                 }
             }
-            Some(Err(e)) => return Err(e),
+            Err(e) => return Err(e),
         }
     }
+    output_handler.flush();
+
     Ok(())
 }
 
-fn u8_to_string(data: &[u8]) -> Cow<'_, str> {
-    String::from_utf8_lossy(data)
+const fn key_count_exceeded(max: Option<u32>, key_count: u32) -> bool {
+    matches!(max, Some(max) if key_count >= max)
+}
+
+fn u8_to_string(data: &[u8]) -> String {
+    String::from_utf8_lossy(data).to_string()
 }
 
 fn key_parser(s: &str) -> Result<Box<[u8]>, std::io::Error> {
     Ok(Box::from(s.as_bytes()))
 }
+
+// Helper function to convert key and value to a string
+fn key_value_to_string(key: &[u8], value: &[u8], hex: bool) -> (String, String) {
+    let key_str = if hex {
+        hex::encode(key)
+    } else {
+        u8_to_string(key)
+    };
+    let value_str = if hex {
+        hex::encode(value)
+    } else {
+        u8_to_string(value)
+    };
+    (key_str, value_str)
+}
+
+async fn handle_next_key(next_key: KeyFromStream) {
+    match next_key {
+        Some(Ok((key, _))) => {
+            println!(
+                "Next key is {0}, resume with \"--start-key={0}\".",
```

Comment:
This could be ugly for non-printable keys, which is the norm. Keys are often hashes, which contain lots of unprintable characters.